docs/proposals/nginx-extensions.md

```diff
@@ -0,0 +1,941 @@
+- [`fail_timeout`](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#server)
+- [`max_fails`](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#server)
+
+NGINX Plus Features:
+
+- Request queueing
+- Slow start
+- Active health checks
+- Load-balancing method (`least_time`)
+
+NGINX Plus directives/parameters:
+
+- [`slow_start`](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#server)
+- [`queue`](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#queue)
+- [`health_check`](https://nginx.org/en/docs/http/ngx_http_upstream_hc_module.html#health_check)
+- [`least_time`](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#ip_hash)
+
+These features are grouped because they all apply to the upstream context and make sense when attached to a Backend.
+
+#### Alternatives
+
+- Split out health checks into a separate Policy: The upstream settings group is large and could be split into two Policies based on use cases. However, we want to limit the number of CRDs we maintain and reduce the overlap between CRDs. Two Policies that set similar directives and parameters in the upstream context could be complex to implement.
+
+### Authentication
+
+_Extension type:_ Inherited Policy
+
+_Resource type:_ CRD
+
+_Role(s):_ Cluster Operator, Application Developer
+
+_Extension point:_ Gateway, Gateway Listener, HTTPRoute Rule (once it's supported)
+
+_NGINX context(s):_ http, server, location
+
+OSS Features:
+
+- Basic Authentication (User/Password)
+- API Key Authentication
+
+OSS NGINX directives:
+
+- [`auth_basic`](https://nginx.org/en/docs/http/ngx_http_auth_basic_module.html#auth_basic)
+- [`auth_basic_user_file`](https://nginx.org/en/docs/http/ngx_http_auth_basic_module.html#auth_basic_user_file)
+
+NGINX Plus Features:
+
+- JSON Web Token (JWT) Authentication
+- OIDC Authentication
+
+NGINX Plus directives:
+
+- [`auth_jwt`](https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt)
+- [`auth_jwt_type`](https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt_type)
+- [`auth_jwt_key_file`](https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt_key_file)
+- [`auth_jwt_key_request`](https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt_key_request)
+
+#### Future Work
+
+- Add support for remote authentication or other authentication strategies.
+
+#### Alternatives
+
+- Begin with a Filter: Rather than beginning with an Inherited Policy, which is complex, we could start by adding an authentication Filter. This would allow us to roll out an authentication solution quickly to support the core use case for Application Developers. The use case for Cluster Operators applying authentication Policies is less well-known than the Application Developer's use case. Instead of leading with the Cluster Operator use case, we can begin with a Filter and wait for user feedback. Later, we can add a Policy and define how the Filter and Policy interact.
```

Comment:
I agree. The downside to this approach is that if we add a Policy for authentication in the future, we will end up with multiple extension types modifying the same setting, which may confuse the user.

Reply:
I see the following benefits for using filters:
(1) 
Inheritance rules can wreak havoc here. 
 
To configure an auth feature, typically multiple fields are involved, and they are not really indepentent. 

For example, let's take an existing NIC JWT Policy -- https://docs.nginx.com/nginx-ingress-controller/configuration/policy-resource/#jwt-using-local-kubernetes-secret

```yaml
jwt:
  secret: jwk-secret
  realm: "My API"
  token: $http_token
```


How, imagine the one above is applied for an HTTPRoute. And now imagine  there is a Gateway policy that overrides the `token` field with some value... the possibility of such overrides can create weird behaviour.

For that case, the inheritance override should be per whole `jwt` field, not its child fields imho. However, I don't think it follows the inheritance policy guidelines.

(2)
with filters, it is clear from reading HTTPRoute that authentication must be applied to It:
```
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    filters:
    - type: ExtensionRef
      extensionRef:
        group: my.group.io
        kind: MyCustomFilter
        name: example
```

More over, if the filter doesn't exist by mistake, NGINX will return 502, so the protected path will not be exposed.

With Policies, by looking at HTTPRoute, you won't see any auth there,
More over if the Policy resource mistakenly doesn't exist, or NGF hasn't reconciled it yet, NGINX will allow traffic to the resource assumably protected :(

Reply:
Good points @pleshakov, I'm convinced. I will add them to the doc, and make the filter the suggested extension type.

Reply:
we'd be curious to here the performance impact, specifically where the hit would come

Reply:
If there's one authentication method for an entire application (maybe SSO), instead of creating a single Policy that applies to all Routes, a user would have to add a Filter to every Route. 

If there are 1000 routes with 10 routing rules each. NGF will process 10,000 references to the same Filter. Versus processing a single Policy that is applied to all 10,000 rules. Of course, applying a single Policy to 10,000 rules may not be any more performant. It's just something we should keep it mind.

Reply:
I see, thanks for clarifying. hard to tell without measuring.

there is an edge case here if the filter becomes invalid, there will be the need to update the corresponding HTTPRoutes (10,000/ 16  (16 is the max path per route I believe)

In my view, I see the real pain here is for users in this case - the need to apply the filter per path.